### PR TITLE
TFLu: Allocate scratch tensors even though they are not in subgraph

### PR DESCRIPTION
**Pull request by mansnils**
This is fixing issue: https://github.com/tensorflow/tensorflow/issues/43537

### CODE REVIEW

**Comment by advaitjain** [comment]
@nkreeger, can you take a look?

File: tensorflow/lite/micro/micro_allocator_test.cc [hunk 1]
```diff
@@ -795,4 +795,71 @@ TF_LITE_MICRO_TEST(TestOperatorInputsNotInSubgraphInputs) {
   TF_LITE_MICRO_EXPECT_EQ(0, eval_tensors[4].data.uint8 - start);
 }
 
+TF_LITE_MICRO_TEST(TestTypicalFirstOpAndSecondOpWithScratchTensors) {
+  constexpr int nbr_tensors = 6;
```
Comment (nkreeger) [inline tensorflow/lite/micro/micro_allocator_test.cc:799]: nit: `number_tensors` - makes this a little easier to read when variables are not abbreviated.

File: tensorflow/lite/micro/micro_allocator_test.cc [hunk 2]
```diff
@@ -749,14 +749,14 @@ TF_LITE_MICRO_TEST(TestOperatorInputsNotInSubgraphInputs) {
...
-
-  int num_conns = 2;
-  tflite::testing::NodeConnection node_list[2] = {
+  constexpr int t0 = 0;
+  constexpr int t1 = 1;
+  constexpr int t2 = 2;
+  constexpr int t3 = 3;
+  constexpr int t4 = 4;
+
+  constexpr int num_conns = 2;
```
Comment (nkreeger) [inline tensorflow/lite/micro/micro_allocator_test.cc:758]: Same comment here - even though this is existing code, would be great if the variable name was spelled out to `number_connections` or similar.

File: tensorflow/lite/micro/micro_allocator_test.cc [hunk 3]
```diff
@@ -795,4 +795,71 @@ TF_LITE_MICRO_TEST(TestOperatorInputsNotInSubgraphInputs) {
   TF_LITE_MICRO_EXPECT_EQ(0, eval_tensors[4].data.uint8 - start);
 }
 
+TF_LITE_MICRO_TEST(TestTypicalFirstOpAndSecondOpWithScratchTensors) {
+  constexpr int nbr_tensors = 6;
+  tflite::AllOpsResolver op_resolver = tflite::testing::GetOpResolver();
+  tflite::NodeAndRegistration* node_and_registration;
+  const int32_t metadata_buffer[tflite::testing::kOfflinePlannerHeaderSize +
+                                nbr_tensors] = {
+      1, 0, nbr_tensors,  // header: version, subgraph, nbr tensors
```
Comment (nkreeger) [inline tensorflow/lite/micro/micro_allocator_test.cc:804]: Google style guide generally suggests having inline comments like this:

```c++
  const int32_t metadata_buffer[tflite::testing::kOfflinePlannerHeaderSize +
                                nbr_tensors] = {/*version=*/1,
                                                /*subgraph=*/0,
                                                number_tensors,
                                                /*t0=*/0,
                                                /*t1=*/0,
                                                /*t2=*/0,
                                                /*t3=*/48,
                                                /*t4=*/-1};
```

**Review by nkreeger — CHANGES_REQUESTED** [review]
Looks mostly good - just some style cleanup.

File: tensorflow/lite/micro/micro_allocator_test.cc (hunk 1 above)
Comment (mansnils) [reply tensorflow/lite/micro/micro_allocator_test.cc:799]: Done

File: tensorflow/lite/micro/micro_allocator_test.cc (hunk 2 above)
Comment (mansnils) [reply tensorflow/lite/micro/micro_allocator_test.cc:758]: Done

File: tensorflow/lite/micro/micro_allocator_test.cc (hunk 3 above)
Comment (mansnils) [reply tensorflow/lite/micro/micro_allocator_test.cc:804]: Done

**Comment by mansnils** [comment]
@nkreeger Ping for review

**Comment by mansnils** [comment]
@advaitjain @nkreeger ping for review

**Review by nkreeger — APPROVED** [review]

**Comment by nkreeger** [comment]
Hi @mansnils this looks like it doesn't build the micro_allocator_test:

```
tensorflow/lite/micro/micro_allocator_test.cc:836:70: error: too few arguments to function call, expected 3, have 2
      kTfLiteOk, allocator->FinishModelAllocation(model, eval_tensors));
                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                    ^
./tensorflow/lite/micro/testing/micro_test.h:118:15: note: expanded from macro 'TF_LITE_MICRO_EXPECT_EQ'
    auto vy = y;                                                               \
              ^
./tensorflow/lite/micro/micro_allocator.h:137:16: note: 'FinishModelAllocation' declared here
  TfLiteStatus FinishModelAllocation(
               ^
1 error generated.
```

**Review by nkreeger — APPROVED** [review]

**Pull request merged**